### PR TITLE
Add support for fsync and fdatasync

### PR DESCRIPTION
**Pull request by vtikoo**
This syscall is nop for ext2 and ramfs based filesystems. For hostfs the
syscall operation is delegated to the host.

For special ramfs based filesystems like devfs and procfs, an override
mechanism is introduced. For fdatasync the override will always fail
with EINVAL.

Signed-off-by: Vikas Tikoo <vitikoo@microsoft.com>

### CODE REVIEW

File: tools/myst/myst.edl [hunk 1]
```diff
@@ -403,6 +403,8 @@ enclave
             gid_t host_egid   
         );
 
+        long myst_fdatasync_ocall(int fd);
```
Comment (jxyang) [inline tools/myst/myst.edl:406]: Can we also implement `fsync` for hostfs in this PR?

File: tests/procfs/procfs.c [hunk 2]
```diff
@@ -99,13 +99,26 @@ int test_cpuinfo()
...
+    int fd;
+    char buf[1024];
+
+    fd = open("/proc/cpuinfo", O_RDONLY);
+    assert(fd > 0);
+    int ret = fdatasync(fd);
+    assert(ret == -1 && errno == EINVAL);
+    close(fd);
+}
+
```
Comment (jxyang) [inline tests/procfs/procfs.c:113]: We should have a test case for hostfs. 
1. stat a host file 
2. Open the host file
3. write to it
4. call fdatasync
5. do stat again, ensure file size has changed.

File: tools/myst/myst.edl (hunk 1 above)
Comment (vtikoo) [reply tools/myst/myst.edl:406]: Added.

File: tests/procfs/procfs.c (hunk 2 above)
Comment (vtikoo) [reply tests/procfs/procfs.c:113]: Added test case on these lines in `/tests/fs`.

File: tests/fs/fs.c [hunk 3]
```diff
@@ -922,6 +922,54 @@ static void test_enotdir()
...
+    assert(fd >= 0);
+    struct stat statbuf;
+    int ret = fstat(fd, &statbuf);
+    assert(ret == 0);
+    int init_size = statbuf.st_size;
+    ret = write(fd, alpha, sizeof(alpha));
+    assert(ret == sizeof(alpha));
+    ret = fdatasync(fd);
+    assert(ret == 0);
+    close(fd);
```
Comment (jxyang) [inline tests/fs/fs.c:937]: I think `close(fd)` triggers data writing back to disk. We should close the file after `fdatasync` is verified.

File: tests/fs/fs.c [hunk 4]
```diff
@@ -922,6 +922,54 @@ static void test_enotdir()
...
+    struct timespec mtime_0 = statbuf.st_mtim;
+
+    // write should change mtime of the file
+    ret = write(fd, alpha, sizeof(alpha));
+    assert(ret == sizeof(alpha));
+
+    // call fsync to flush file and its metadata to device
+    ret = fsync(fd);
+    assert(ret == 0);
+    close(fd);
```
Comment (jxyang) [inline tests/fs/fs.c:963]: Likewise.

File: tests/fs/fs.c (hunk 3 above)
Comment (vtikoo) [reply tests/fs/fs.c]: Fixed.

File: tests/fs/fs.c (hunk 4 above)
Comment (vtikoo) [reply tests/fs/fs.c]: Fixed.

**Review by jxyang — APPROVED** [review]

**Comment by vtikoo** [comment]
Failing pipeline test is unrelated to this PR.

**Pull request merged**